core/state-db/src/lib.rs

```diff
@@ -200,18 +221,41 @@ impl<BlockHash: Hash, Key: Hash> StateDbSync<BlockHash, Key> {
 		})
 	}
 
+	fn check_meta<D: MetaDb>(mode: &PruningMode, db: &D) -> Result<(), Error<D::Error>> {
+		let db_mode = db.get_meta(&to_meta_key(PRUNING_MODE, &())).map_err(Error::Db)?;
+		trace!(target: "state-db",
+			"DB pruning mode: {:?}",
+			db_mode.as_ref().map(|v| std::str::from_utf8(&v))
+		);
+		match &db_mode {
+			Some(v) if v.as_slice() == mode.id() => Ok(()),
+			Some(v) => Err(Error::InvalidPruningMode(String::from_utf8_lossy(v).into())),
+			None => Ok(()),
+		}
+	}
+
 	pub fn insert_block<E: fmt::Debug>(&mut self, hash: &BlockHash, number: u64, parent_hash: &BlockHash, mut changeset: ChangeSet<Key>) -> Result<CommitSet<Key>, Error<E>> {
+		let mut meta = ChangeSet::default();
+		if number == 0 {
+			// Save pruning mode when writing first block.
+			meta.inserted.push((to_meta_key(PRUNING_MODE, &()), self.mode.id().into()));
```

Comment:
Can't we do this above on `new`? When checking meta if it comes back as `None` then we update it.

Reply:
`new` and `StateDB` in general does not write to the database directly.  It returns a `CommitSet` that is applied in the upper level code. Returning a `CommitSet` from `new` seems less elegant and would also require some restructuring of the calling code.